lib/mayaUsd/ufe/Utils.cpp

```diff
@@ -607,79 +610,110 @@ Ufe::Attribute::Type usdTypeToUfe(const SdfValueTypeName& usdType)
-            Ufe::Attribute::Type type = usdTypeToUfe(it->second);
-            if (type == Ufe::Attribute::kString && !shaderProperty->GetOptions().empty()) {
-                return Ufe::Attribute::kEnumString;
-            } else {
-                return type;
-            }
+            retVal = _UsdTypeToUfe(it->second);
         } else {
 #if PXR_VERSION < 2205
             // Pre-22.05 boolean inputs are special:
             if (shaderProperty->GetType() == SdfValueTypeNames->Bool.GetAsToken()) {
-                return usdTypeToUfe(PXR_NS::SdfValueTypeNames->Bool);
-            }
+                retVal = _UsdTypeToUfe(SdfValueTypeNames->Bool);
+            } else
 #endif
-            // There is no Matrix3d type in Sdr, so we need to infer it from Sdf until a fix similar
-            // to what was done to booleans is submitted to USD. This also means that there will be
-            // no default value for that type.
-            if (shaderProperty->GetType() == SdfValueTypeNames->Matrix3d.GetAsToken()) {
-                return usdTypeToUfe(PXR_NS::SdfValueTypeNames->Matrix3d);
+                // There is no Matrix3d type in Sdr, so we need to infer it from Sdf until a fix
+                // similar to what was done to booleans is submitted to USD. This also means that
+                // there will be no default value for that type.
+                if (shaderProperty->GetType() == SdfValueTypeNames->Matrix3d.GetAsToken()) {
+                retVal = _UsdTypeToUfe(SdfValueTypeNames->Matrix3d);
+            } else {
+                retVal = Ufe::Attribute::kGeneric;
             }
-            return usdTypeToUfe(PXR_NS::SdfValueTypeNames->Token);
         }
     } else {
-        Ufe::Attribute::Type type = usdTypeToUfe(typeName);
-        if (type == Ufe::Attribute::kString && !shaderProperty->GetOptions().empty()) {
-            return Ufe::Attribute::kEnumString;
-        } else {
-            return type;
+        retVal = _UsdTypeToUfe(typeName);
+    }
+
+    if (retVal == Ufe::Attribute::kString && !shaderProperty->GetOptions().empty()) {
+        retVal = Ufe::Attribute::kEnumString;
+    }
+
+    return retVal;
+}
+
+Ufe::Attribute::Type usdTypeToUfe(const PXR_NS::UsdAttribute& usdAttr)
+{
+    if (usdAttr.IsValid()) {
+        const SdfValueTypeName typeName = usdAttr.GetTypeName();
+        Ufe::Attribute::Type   type = _UsdTypeToUfe(typeName);
+        if (type == Ufe::Attribute::kString) {
+            // Both std::string and TfToken resolve to kString, but if there is a list of allowed
+            // tokens, then we use kEnumString instead.
+            auto attrDefn
+                = usdAttr.GetPrim().GetPrimDefinition().GetSchemaAttributeSpec(usdAttr.GetName());
+            if (attrDefn && attrDefn->HasAllowedTokens()) {
+                type = Ufe::Attribute::kEnumString;
+            }
+            // TfToken is also used in UsdShade as a Generic placeholder for connecting struct I/O.
```

Comment:
Necessary because we want to have generic input and output ports on NodeDef boundaries and this requires getting them back as Generic even though they have TfToken USD data type.